streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java

```diff
@@ -1040,283 +1042,91 @@ <KOut> KGroupedStream<KOut, V> groupBy(final KeyValueMapper<? super K, ? super V
-     * Furthermore, both input streams need to be co-partitioned on the join key (i.e., use the same partitioner).
-     * If this requirement is not met, Kafka Streams will automatically repartition the data, i.e., it will create an
-     * internal repartitioning topic in Kafka and write and re-read the data via this topic before the actual join.
-     * The repartitioning topic will be named "${applicationId}-&lt;name&gt;-repartition", where "applicationId" is
-     * user-specified in {@link StreamsConfig} via parameter
-     * {@link StreamsConfig#APPLICATION_ID_CONFIG APPLICATION_ID_CONFIG}, "&lt;name&gt;" is an internally generated
-     * name, and "-repartition" is a fixed suffix.
-     * <p>
-     * Repartitioning can happen for one or both of the joining {@code KStream}s.
-     * For this case, all data of the stream will be redistributed through the repartitioning topic by writing all
-     * records to it, and rereading all records from it, such that the join input {@code KStream} is partitioned
-     * correctly on its key.
-     * <p>
-     * Both of the joining {@code KStream}s will be materialized in local state stores with auto-generated store names.
+     * If this is not the case (and if not auto-repartitioning happens, see further below),
+     * you would need to call {@link #repartition(Repartitioned)} (for at least one of both
+     * {@code KStreams}) before doing the join and specify the "correct" number of partitions via {@link Repartitioned}
+     * parameter to align the partition count for both inputs to each other.
+     * Furthermore, both {@code KStreams} need to be co-partitioned on the join key (i.e., use the same partitioner).
+     * Note: Kafka Streams cannot verify the used partitioning strategy, so it is the user's responsibility to ensure
+     * that the same partitioner is used for both inputs for the join.
+     *
+     * <p>If a key changing operator was used before this operation on either input stream
+     * (e.g., {@link #selectKey(KeyValueMapper)}, {@link #map(KeyValueMapper)}, {@link #flatMap(KeyValueMapper)} or
+     * {@link #process(ProcessorSupplier, String...)}) Kafka Streams will automatically repartition the data of the
+     * corresponding input stream, i.e., it will create an internal repartitioning topic in Kafka and write and re-read
+     * the data via this topic such that data is correctly partitioned by the join key.
+     *
+     * <p>The repartitioning topic(s) will be named "${applicationId}-&lt;name&gt;-repartition",
+     * where "applicationId" is user-specified in {@link StreamsConfig} via parameter
+     * {@link StreamsConfig#APPLICATION_ID_CONFIG APPLICATION_ID_CONFIG},
+     * "&lt;name&gt;" is an internally generated name, and "-repartition" is a fixed suffix.
+     * The number of partitions for the repartition topic(s) is determined based on both upstream topics partition
```

Comment:
```suggestion
     * The number of partitions for the repartition topic(s) is determined based on both upstream topics' partition
```

Not even sure, but `upstream topics partition numbers` is a bit confusing. Maybe `...based on the partition numbers of both upstream topics`.